### ISSUE TITLE
[Question] Switch between constants

### ISSUE THREAD

**Issue by TimSchimansky** [kaitai-io/kaitai_struct#962]
Hi,
tldr: I am searching for a method to switch between 64 constant values to use for multiplication in an instance value expression.

I have the data of a laserscanner (Hesai Pandar 64). The scanner spits out a UDP package with header, body and tail. The body consists of 1164 bytes and is dividable into 6 blocks. Each blocks contains one 2 bit integer for the azimuth (horizontal angle) value and 3 bytes (2 for distance, 1 for reflectance) for each of the 64 laser beams.

The readout of this works fine and maybe the goal to archieve this with kaitai is overambitious... I have a calibration file for the Pandar64, where every of the 64 channels gets an elevation angle (as well as an azimuth offset; same thing but i only need to describe one thing).  My goal is to convert the polar coordinates (azimuth/elevation) into cartesian coordinates (x/y/z). Therefore i need kind of a lookup table for the values (i have tried it with enums but ran into the problem of not accepting numbervalues as values). Is this even feesable.

In the following example i dont even try to calculate anything:

```
[...]
  
seq:
    - id: header
      type: header
    - id: blocks
      type: block
      repeat: expr
      repeat-expr: header.n_blocks
    - id: tail
      type: tail
    
types:
    
    [...]
	
    block:
        seq:
            - id: azimuth_raw
              type: u2
              doc: divide by 100 degrees
            - id: channel
              type: channel(_index)
              repeat: expr
              repeat-expr: _parent.header.n_lasers
        instances:
            azimuth_deg:
                value: azimuth_raw.as<f8> / 100
              
    [...]

    channel:
        params:
            - id: i
              type: u1
        seq:
            - id: distance_value_raw
              type: u2
              doc: Multiplication with distance_unit needed
            - id: reflectance_value
              type: u1
              doc: Reflectance on a scale of 0 to 255
        instances:
            distance_value:
                value: distance_value_raw * _parent._parent.header.distance_unit
				
			# THIS IS THE PROBLEM #####################
            elevation:
                value: i
                enum: elevations

enums:
    elevations:
        1: 14.803
        2: 10.953
        3: 7.98
      
		[...]
	  
        61: -13.053
        62: -14.009
        63: -18.968
        64: -24.976
```

**Comment by KOLANICH** [comment]
1. `enums` map integers to symbolic names just for convenience
2. My best advice is to try to figure out the mapping in a form of an analytic formula using math.
3. My next best advice is to just mention it in `doc`, if it is infeasible to put it into KS (i.e. KS lacks trigonometric functions and even powers, for powers I used a script generating square-and-multiply reimpl)
4. If one cannot figure out the folrmula, one can always use a look-up table in a form of array of floats. But AFAIK you cannot declare it right in KS.
5. So let's assumme you cannot do this in pure KS, but want to have the lookup right in the spec. You have options. You can create an opaque type and implement the opaque type yourself to have a look-up table. Then you can try to look-up it. Not sure though it'll work for dynamic languages. Another option is to generate th3 code without lookup and with it (but looking up the wrong place, the one KSC cangenerate), modify the code with lookup the way it used the correct lookup, and generate a patch file in a form of a unified diff. The best way to do it is to create an empty repo, commit there a ksc generated file, then commit there the modified file, then `git format-patch -1`. Then apply this patch to the KSC-generated file `patch -p1 < your.patch`.

**Comment by generalmimon** [comment]
@KOLANICH:

> 4\. If one cannot figure out the folrmula, one can always use a look-up table in a form of array of floats. But AFAIK you cannot declare it right in KS.

Why not?

```ksy
meta:
  id: t
instances:
  elevation_lookup_table:
    value: |
      [
        0, 14.803, 10.953, 7.98, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, -13.053, -14.009, -18.968, -24.976,
      ]
  elevation_2:
    value: elevation_lookup_table[2]
  elevation_62:
    value: elevation_lookup_table[62]
```

This is a perfect use case for a lookup table.

**Comment by TimSchimansky** [comment]
hm okay. Thanks for your answers. The lack of trigonometric functions seems to be the issue then...

**Comment by KOLANICH** [comment]
>Why not?

Thanks. Nice to know it exists.
